edx_proctoring/api.py

```diff
@@ -919,20 +904,138 @@ def _check_eligibility_of_enrollment_mode(credit_state):
+                                   filter_out_namespaces=None):
+    """
+    Returns a tuple about the fulfillment of any pre-requisites in order to this exam
+    as proctored. The pre-requisites are taken from the credit requirements table. So if ordering
+    of requirements are - say - ICRV1, Proctoring1, ICRV2, and Proctoring2, then the user cannot take
+    Proctoring2 until there is an explicit pass on ICRV1, Proctoring1, ICRV2...
+
+    NOTE: If evaluate_for_requirement_name=None that means check all requirements
+
+    Return (dict):
+
+        {
+            # If all prerequisites are satisfied
+            'are_prerequisites_satisifed': True/False,
+
+            # A list of prerequisites that have been satisfied
+            'satisfied_prerequisites': [...]
+
+            # A list of prerequisites that have failed
+            'failed_prerequisites': [....],
+
+            # A list of prerequisites that are still pending
+            'pending_prerequisites': [...],
+        }
+
+    NOTE: We filter out any 'grade' requirement since the student will most likely not have fulfilled
+    those requirements while he/she is in the course
+    """
+
+    satisfied_prerequisites = []
+    failed_prerequisites = []
+    pending_prerequisites = []
+
+    # insure an ordered and filtered list
+    # we remove 'grade' requirements since those cannot be
+    # satisfied while student is in the middle of a course
+    requirement_statuses = _get_ordered_prerequisites(
+        prerequisites_statuses,
+        filter_out_namespaces=filter_out_namespaces
+    )
+
+    # find ourselves in the list
+    ourself = None
+    if evaluate_for_requirement_name:
+        for idx, requirement in enumerate(requirement_statuses):
+            if requirement['name'] == evaluate_for_requirement_name:
+                ourself = requirement
+                break
+
+    # we are not in the list of requirements, look at all requirements
+    if not ourself:
+        idx = len(requirement_statuses)
+
+    # now that we have the index of ourselves in the ordered list, we can walk backwards
+    # and inspect all prerequisites
+
+    # we don't look at ourselves
+    idx = idx - 1
+    while idx >= 0:
+        requirement = requirement_statuses[idx]
+        status = requirement['status']
+        if status == 'satisfied':
+            satisfied_prerequisites.append(requirement)
+        elif status == 'failed':
+            failed_prerequisites.append(requirement)
+        else:
+            pending_prerequisites.append(requirement)
+
+        idx = idx - 1
+
+    return {
+        # all prequisites are satisfied if there are no failed or pending requirement
+        # statuses
+        'are_prerequisites_satisifed': (
+            not failed_prerequisites and not pending_prerequisites
+        ),
+        # note that we reverse the list here, because we assempled it by walking backwards
+        'satisfied_prerequisites': list(reversed(satisfied_prerequisites)),
```

Comment:
`sorted(satisfied_prerequisites, reverse=True)` might be better here since reversed returns a iterator and sorted returns a list.